### ISSUE TITLE
[Bug] dbt deps --project-dir $DBT_PROJECT_DIR hangs with dbt=1.5.0

### ISSUE THREAD

**Issue by clausherther** [dbt-labs/dbt-core#7491]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

After upgrading dbt-core to 1.5.0, our CI jobs hangs after running the following command from the parent directory of `integration_tests`:
```
dbt deps --project-dir integration_tests
```
Output:
```
#!/bin/bash -eo pipefail
. venv/bin/activate
  dbt deps --project-dir $DBT_PROJECT_DIR
01:29:08  Running with dbt=1.5.0
```
This will eventually result in a timeout [error](https://app.circleci.com/pipelines/github/calogica/dbt-expectations/253/workflows/92c6f533-2054-4b39-b033-ca8d77701187/jobs/177/parallel-runs/0/steps/0-103).

This was previously working under dbt version 1.4.1 and is currently blocking CI jobs for `dbt-expectations` and `dbt-date`.

If you `cd integrations_tests` and then run `dbt deps`, this command succeeds:

```
integration_tests git:(main) dbt deps
01:37:38  Running with dbt=1.5.0
01:37:39  Installing ../
01:37:39  Installed from <local @ ../>
01:37:39  Installing calogica/dbt_date
01:37:39  Installed from version 0.7.2
01:37:39  Up to date!
``` 

### Expected Behavior

This was previously working under dbt version 1.4.1 and should simply install the packages defined in `integration_tests/packages.yml`.

### Steps To Reproduce

Using dbt=1.5.0, in a project that contains a subproject, such as dbt-utils or dbt-expectations, run the following from the root folder:

`dbt deps --project-dir subproject` where `subproject` is the name of the folder containing the subproject.

### Environment

```markdown
- OS: MacOS 13.3.1
- Python: 3.9
- dbt: 1.5.0
```

### Which database adapter are you using with dbt?

postgres

**Comment by dbeatty10** [comment]
Thank you for reporting this @clausherther 🙏 

I was able to reproduce what you described between 1.4 (working) and 1.5 (terminating after 10 minutes), and I'm going to label this as a regression.

### Reprex

Directory structure:
```
.
├── dbt_project.yml
├── packages.yml
└── subproject
    ├── dbt_project.yml
    └── packages.yml
```

`dbt_project.yml`
```yaml
name: "my_dbt_project"
version: "1.0.0"
config-version: 2
profile: "sandcastle"
```

`packages.yml`
```yaml
packages:
  - package: calogica/dbt_date
    version: [">=0.7.0", "<0.8.0"]
```

`subproject/dbt_project.yml`
```yaml
name: "my_dbt_subproject"
version: "1.0.0"
config-version: 2
profile: "sandcastle"
```

`subproject/packages.yml`
```yaml
packages:
    - local: ../
```

Install dependencies:
```shell
 dbt deps --project-dir subproject
 ```

**Comment by dbeatty10** [comment]
### TLDR
This problem is triggered when both `packages.yml` _and_ `--profiles-dir` contain relative paths.

### More detail

Some more findings from troubleshooting:
- Affects both `--project-dir` and `$DBT_PROJECT_DIR`
- Absolute paths work fine for both
- Only affects relative paths

This works in a `zsh` shell:
```shell
dbt deps --project-dir $(pwd)/subproject
```

Also, when `subproject/packages.yml` contains absolute paths, it works regardless if `--profiles-dir` is relative or not.

**Comment by dbeatty10** [comment]
@iknox-fa I suspect there is some kind of circular dependency that gets introduced in the scenario outlined above (which maybe introduces an infinite loop 🤷).
